sds_in_a_box/sds_in_a_box_stack.py

```diff
@@ -84,56 +84,85 @@ def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
-                                            "OS_DOMAIN": domain.domain_endpoint,
+                                            "OS_DOMAIN": sds_metadata_domain.domain_endpoint,
                                             "OS_PORT": "443",
                                             "OS_INDEX": "metadata",
                                             "S3_BUCKET": data_bucket.s3_url_for_object()}
                                           )
-
         indexer_lambda.add_event_source(S3EventSource(data_bucket,
                                                       events=[s3.EventType.OBJECT_CREATED]
                                                       )
                                         )
         indexer_lambda.apply_removal_policy(cdk.RemovalPolicy.DESTROY)
 
+        # Adding Opensearch permissions 
+        indexer_lambda.add_to_role_policy(
+            iam.PolicyStatement(
+                effect=iam.Effect.ALLOW,
+                actions=["es:*"],
+                resources=[f"{sds_metadata_domain.domain_arn}/*"],
+            )
+        )
+        # Adding S3 Permissions 
+        indexer_lambda.add_to_role_policy(
+            iam.PolicyStatement(
+                effect=iam.Effect.ALLOW,
+                actions=["s3:*"],
+                resources=[
+                    f"{data_bucket.bucket_arn}/*"
+                ],
+            )
+        )
+
         # The purpose of this lambda function is to trigger off of a lambda URL.
         query_api_lambda = lambda_alpha_.PythonFunction(self,
                                           id="QueryAPILambda",
+                                          function_name='download-query-api',
                                           entry=os.path.join(os.path.dirname(os.path.realpath(__file__)), "SDSCode/"),
                                           index="queries.py",
                                           handler="lambda_handler",
-                                          role=lambda_role,
                                           runtime=lambda_.Runtime.PYTHON_3_9,
                                           timeout=cdk.Duration.minutes(1),
                                           memory_size=1000,
                                           environment={
                                             "OS_ADMIN_USERNAME": "master-user", 
                                             "OS_ADMIN_PASSWORD_LOCATION": os_secret.secret_value.unsafe_unwrap(),
-                                            "OS_DOMAIN": domain.domain_endpoint,
+                                            "OS_DOMAIN": sds_metadata_domain.domain_endpoint,
                                             "OS_PORT": "443",
                                             "OS_INDEX": "metadata"
                                             }
                                           )
+        query_api_lambda.add_to_role_policy(
+            iam.PolicyStatement(
```

Comment:
Are all these lambdas expected to have some of the same permissions? It might be worth putting these `iam.PolicyStatement` values into one variable, so you only need to update in one place.

Reply:
Haha yeah that's true, they should all be pretty much the same so it makes sense to put them into variables.  The only potential benefit I could see to do it this way is if each lambda would have unique policies, which they might depending on how deep we want to go with least-privilege restrictions.    

For instance, we could find out the exact actions the query API needs to perform on opensearch from this list: https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazonopensearchservice.html, and rather than give it permissions to do all those actions to our opensearch cluster ("es:*"), we determine only the specific actions we need to grant, such as "es:DescribeElasticsearchDomain", "es:ESHttpDelete", "es:ESHttpHead", etc

Reply:
Definitely, if the lambdas are expected to have different permissions, we should plan to have them only have what's needed. Then, my comment doesn't really apply, because it's mostly placeholders anyway.